p2p/base_reactor.go

```diff
@@ -66,20 +81,138 @@ type EnvelopeReceiver interface {
 //--------------------------------------
 
 type BaseReactor struct {
+	sync.Mutex
 	service.BaseService // Provides Start, Stop, .Quit
 	Switch              *Switch
+
+	incoming chan UnprocessedEnvelope
+
+	// processor is called with the incoming channel and is responsible for
+	// unmarshalling the messages and calling Receive on the reactor.
+	processor func(incoming <-chan UnprocessedEnvelope) error
 }
 
-func NewBaseReactor(name string, impl Reactor) *BaseReactor {
-	return &BaseReactor{
+type ReactorOptions func(*BaseReactor)
+
+func NewBaseReactor(name string, impl Reactor, opts ...ReactorOptions) *BaseReactor {
+	base := &BaseReactor{
 		BaseService: *service.NewBaseService(nil, name, impl),
 		Switch:      nil,
+		incoming:    make(chan UnprocessedEnvelope, 100),
+		processor:   DefaultProcessor(impl),
+	}
+	for _, opt := range opts {
+		opt(base)
+	}
+
+	go func() {
+		err := base.processor(base.incoming)
+		if err != nil {
+			err = base.Stop()
+			if err != nil {
+				panic(err)
+			}
+		}
+	}()
+
+	return base
+}
+
+// WithProcessor sets the processor function for the reactor. The processor
+// function is called with the incoming channel and is responsible for
+// unmarshalling the messages and calling Receive on the reactor.
+func WithProcessor(processor func(<-chan UnprocessedEnvelope) error) ReactorOptions {
+	return func(br *BaseReactor) {
+		br.processor = processor
+	}
+}
+
+// WithIncomingQueueSize sets the size of the incoming message queue for a
+// reactor.
+func WithIncomingQueueSize(size int) ReactorOptions {
+	return func(br *BaseReactor) {
+		br.incoming = make(chan UnprocessedEnvelope, size)
 	}
 }
 
 func (br *BaseReactor) SetSwitch(sw *Switch) {
 	br.Switch = sw
 }
+
+// QueueUnprocessedEnvelope is called by the switch when an unprocessed
+// envelope is received. Unprocessed envelopes are immediately buffered in a
+// queue to avoid blocking. The size of the queue can be changed by passing
+// options to the base reactor.
+func (br *BaseReactor) QueueUnprocessedEnvelope(e UnprocessedEnvelope) {
+	br.Lock()
+	defer br.Unlock()
```

Comment:
why do we need to lock before putting something in a channel?

Reply:
It's a weird data race that happens when we close the channel and try to add items to it.

I'll use a context instead. This was just to confirm the data race source

Reply:
ahh weird

Reply:
809a8cf4235034b00f2bc93eca0d897f8cf8c43a